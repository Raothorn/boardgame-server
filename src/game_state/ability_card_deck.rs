use serde::Serialize;

#[derive(Clone, Serialize, Default)]
pub struct AbilityCard {
    name: String,
    deck_ix: u32,

    modifiers: Vec<Modifier>,
}

impl AbilityCard {
    fn new(name: &str, deck_ix: u32) -> Self {
        AbilityCard {
            name: name.to_owned(),
            deck_ix,
            modifiers: Vec::new(),
        }
    }
}

pub fn ability_card_deck() -> Vec<AbilityCard> {
    let card1 = AbilityCard::new("Triage", 1);
    let card2 = AbilityCard::new("Focused Mind", 2);
    let card3 = AbilityCard::new("Counsel", 3);

    vec![card1, card2, card3]
}

type Modifier = ();
