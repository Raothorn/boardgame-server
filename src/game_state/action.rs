use core::fmt;
use std::fmt::Display;

use super::{GameState, Update};
use serde::{Deserialize, Serialize};
use serde_json::{Value, json};

mod accept_challenge_result_action;
mod accept_message_action;
mod choose_token_for_deck_action;
mod draw_for_deck_action;
mod handle_event_phase_action;
mod resolve_challenge_action;
mod select_discard_for_galley_action;
mod select_event_option_action;
mod take_ship_action;
mod select_main_action;
mod travel_action;
mod equip_ability_card_action;
mod select_crew_member_action;

#[typetag::serde(tag = "actionType", content = "actionData")]
pub trait Action: fmt::Display {
    fn execute(&self, state: &GameState) -> Update<GameState> {
        Ok(state.clone())
    }
}

pub fn get_action(action_msg_str: &str) -> Box<dyn Action>{
    let msg: Value = serde_json::from_str(action_msg_str).unwrap();

    let result = serde_json::from_value::<Box<dyn Action>>(msg);
    match result {
        Ok(action) => action,
        
        // We want to panic if an action isn't being parsed
        Err(err) => panic!("{}", err),
    }
}


// BASIC ACTIONS

#[derive(Deserialize, Serialize, Clone)]
struct NoAction;

#[typetag::serde(name = "noAction")]
impl Action for NoAction {
    fn execute(&self, gs: &GameState) -> Update<GameState> {
        Ok(gs.to_owned())
    }
}

impl Display for NoAction {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "No action")
    }
}
