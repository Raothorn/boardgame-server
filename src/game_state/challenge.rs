use serde::Serialize;

use super::{skill::Skill, Update, GameState};

#[derive(Clone, Serialize)]
pub struct Challenge {
    pub skill: Skill,
    pub amount: u32,

    #[serde(skip_serializing)]
    pub if_fail: fn(&GameState) -> Update,
    #[serde(skip_serializing)]
    pub if_succeed: fn(&GameState) -> Update
}

impl Default for Challenge {
    fn default() -> Self {
        Self {
            skill: Skill::Craft,
            amount: Default::default(),
            if_fail: |gs| Ok(gs.clone()),
            if_succeed: |gs| Ok(gs.clone())
        }
    }
}
