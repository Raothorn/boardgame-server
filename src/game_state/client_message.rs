use serde::Serialize;

use super::AbilityCard;

#[derive(Clone, Serialize)]
pub enum ClientMessage {
    GainCommandPoints { amount: u32 },
    DrewAbilityCard { card: AbilityCard }
}
