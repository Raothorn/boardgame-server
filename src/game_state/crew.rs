use serde::Serialize;
use serde_with::DisplayFromStr;

use super::skill::Skill;
use std::collections::HashMap;

#[serde_with::serde_as]
#[derive(Serialize, Clone)]
pub struct Crew {
    pub name: String,
    pub fatigue: u32,
    #[serde_as(as = "HashMap<DisplayFromStr,_>")]
    pub skills: HashMap<Skill, u32>,
}

impl Crew {
    pub fn new(name: &str, savvy: u32, craft: u32) -> Self {
        let mut skills = HashMap::new();
        skills.insert(Skill::Savvy, savvy);
        skills.insert(Skill::Craft, craft);

        Crew { name: name.to_owned(), fatigue:0, skills }
    }

    pub fn reduce_fatigue(&mut self) {
        if self.fatigue > 0 {
            self.fatigue -= 1;
        }
    }
}
