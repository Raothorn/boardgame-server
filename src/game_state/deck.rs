use serde::Serialize;

#[derive(Serialize, Clone)]
pub struct Deck<T: Clone> {
    items: Vec<T>,
    discard: Vec<T>,
}

impl<T: Clone> Deck<T> {
    pub fn new(items: Vec<T>) -> Self {
        Deck {
            items,
            discard: Vec::new(),
        }
    }

    pub fn draw(self) -> Result<(Self, T), String> {
        let mut deck = self.clone();
        if deck.items.is_empty() {
            deck.items.append(&mut deck.discard);
            println!("none left");
        }

        deck.items
            .pop()
            .ok_or("No items left in the deck".to_string())
            .map(|item| (deck, item))
    }

    pub fn add_to_discard(&mut self, item: &T) {
        self.discard.push(item.clone());
    }
}
