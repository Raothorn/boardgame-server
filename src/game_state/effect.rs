use std::fmt::Display;

use serde::{Deserialize, Serialize};
use serde_json::json;

use super::{
    action::Action,
    challenge::Challenge,
    crew::Status,
    game_phase::{GamePhase, MainActionSubphase},
    GameState, Resource, Update,
};

#[derive(Clone, Serialize, Debug)]
pub enum Effect {
    TryChallenge(Challenge),
    GainResource(Resource, i32),
    TakeHealthDamage(u32),
    TakeStatus(Status),
    TurnToStory(String),
    ReturnToShip,
}

impl Effect {
    pub fn resolve(&self, state: GameState) -> Update<GameState> {
        match self {
            Effect::TurnToStory(story_ix) => {
                state.map.storybook.story(story_ix).clone().and_then(
                    |story| {
                        state.set_phase(GamePhase::ExplorePhase(
                            story.clone(),
                        ))
                    },
                )
            }

            Effect::ReturnToShip => state.pop_phase().and_then(|g| {
                if let GamePhase::MainActionPhase(
                    Some(MainActionSubphase::Explore),
                    action_ct,
                ) = g.phase()
                {
                    let phase =
                        GamePhase::MainActionPhase(None, action_ct);

                    g.set_phase(phase)
                } else {
                    Err("Wrong phase".to_owned())
                }
            }),

            Effect::GainResource(resource, amount) => {
                state.gain_resource(*resource, *amount)
            }

            Effect::TryChallenge(challenge) => {
                state.challenge(challenge.clone())
            }

            Effect::TakeStatus(status) => {
                let action = TakeStatusAction { status: *status };
                let action_ser = json! ({
                    "actionType": "takeStatusAction",
                    "actionData": action
                })
                .to_string();
                let display = format!("Give {:?} to", status);

                state.push_phase(GamePhase::SelectCrewMemberPhase {
                    crew_ix: None,
                    title: display,
                    callback: action_ser,
                })
            }

            _ => state.push_phase(GamePhase::ResolveEffectPhase(
                self.clone(),
            )),
        }
    }
}

pub fn resolve_effects(
    state: &GameState,
    effects: Vec<Effect>,
) -> Update<GameState> {
    effects.into_iter().fold(Ok(state.clone()), |g, eff| {
        g.and_then(|g| eff.resolve(g))
    })
}

#[derive(Clone, Serialize, Deserialize)]
pub struct TakeStatusAction {
    status: Status,
}

#[typetag::serde(name = "takeStatusAction")]
impl Action for TakeStatusAction {
    fn execute(&self, state: &GameState) -> Update<GameState> {
        // TODO add string to phase
        if let GamePhase::SelectCrewMemberPhase{crew_ix: Some(crew_ix), title: _, callback: _} =
            state.phase()
        {
            state
                .clone()
                .update_crew(crew_ix, |c| c.add_status(self.status))
        } else {
            Err("wrong phase".to_owned())
        }
    }
}

impl Display for TakeStatusAction {
    fn fmt(
        &self,
        f: &mut std::fmt::Formatter<'_>,
    ) -> std::fmt::Result {
        write!(f, "Take Status")
    }
}
