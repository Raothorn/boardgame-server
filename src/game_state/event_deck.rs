use serde::Serialize;

use crate::game_state::{challenge::Challenge, Skill};

use super::{GameState, Update};

#[derive(Clone, Serialize)]
pub struct EventCard {
    pub name: String,
    pub options: Vec<EventOption>,
    // TODO: deck index
}

#[derive(Clone, Serialize)]
pub struct EventOption {
    pub text: String,
    #[serde(skip_serializing)]
    pub handle_option: fn(&GameState) -> Update,
}

pub fn event_deck() -> Vec<EventCard> {
    vec![
        EventCard {
            name: "Biting Starfish".to_owned(),
            options: vec![
                EventOption {
                    text: "Gain 1 Coin".to_owned(),
                    handle_option: gain_coin,
                },
                EventOption {
                    text: "Gain 1 Meat".to_owned(),
                    handle_option: gain_meat,
                },
            ],
        },
        EventCard {
            name: "Broken Biplane".to_owned(),
            options: vec![EventOption {
                text: "Help repair the airplane: CRAFT 8".to_owned(),
                handle_option: (|g| {
                    g.challenge(Challenge {
                        skill: Skill::Craft,
                        amount: 3,
                    })
                }),
            }],
        },
    ]
}

fn gain_coin(state: &GameState) -> Update {
    let mut gs = state.clone();
    gs.resources.coins += 1;

    Ok(gs)
}

fn gain_meat(state: &GameState) -> Update {
    let mut gs = state.clone();
    gs.resources.meat += 1;

    Ok(gs)
}
