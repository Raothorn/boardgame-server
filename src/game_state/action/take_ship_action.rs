use std::fmt::Display;

use serde::Deserialize;

use super::Action;

use crate::game_state::{
    game_phase::ShipActionSubphase, GamePhase, GameState, ShipRoom,
    Update,
};

#[derive(Deserialize)]
pub struct TakeShipAction {
    room: ShipRoom,
    player_ix: usize,
}

impl TakeShipAction {
    fn bridge_action(&self, state: &GameState) -> Update {
        Ok(state.clone())
            .and_then(|g| g.give_command_tokens(self.player_ix, 3))
            .and_then(|g| g.draw_cards(self.player_ix, 1))
            .map(|g| GameState {
                room: ShipRoom::Bridge,
                ..g
            })
            .and_then(|g| g.set_phase(GamePhase::EventPhase(None)))
    }

    fn deck_action(&self, state: &GameState) -> Update {
        let phase = GamePhase::ShipActionPhase(Some(
            ShipActionSubphase::DeckAction {
                search_tokens_drawn: Vec::new(),
            },
        ));
        Ok(state.clone())
            .map(|g| GameState {
                room: ShipRoom::Deck,
                ..g
            })
            .and_then(|g| g.set_phase(phase))
    }

    fn galley_action(&self, state: &GameState) -> Update {
        let phase = GamePhase::ShipActionPhase(Some(
            ShipActionSubphase::GalleyAction
        ));

        Ok(state.clone())
            .and_then(|g| g.give_command_tokens(self.player_ix, 3))
            .and_then(|g| g.draw_cards(self.player_ix, 2))
            .and_then(|g| g.set_room(&ShipRoom::Galley))
            .and_then(|g| g.set_phase(phase))
    }
}

impl Action for TakeShipAction {
    fn execute(&self, state: &GameState) -> Update {
        if let GamePhase::ShipActionPhase(None) = &state.phase() {
            if state.room == self.room {
                Err("You cannot visit the same room twice".to_owned())
            } else {
                match self.room {
                    ShipRoom::Bridge => self.bridge_action(state),
                    ShipRoom::Galley => self.galley_action(state),
                    ShipRoom::Deck => self.deck_action(state),
                    _ => Err(String::from("Not implemented")),
                }
            }
        } else {
            Err("Wrong phase.".to_owned())
        }
    }
}

impl Display for TakeShipAction {
    fn fmt(
        &self,
        f: &mut std::fmt::Formatter<'_>,
    ) -> std::fmt::Result {
        write!(
            f,
            "Take Ship Action:\n Room: {:?}\n PlayerIx: {}",
            self.room, self.player_ix
        )
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use test_case::test_case;

    use crate::game_state::challenge::Challenge;
    use GamePhase::ChallengePhase as Cp;
    use GamePhase::EventPhase as Ep;
    use GamePhase::ShipActionPhase as Sa;
    use ShipActionSubphase as Sas;

    #[test]
    fn test_bridge_action() {
        let gs = GameState::init_state();
        let action = TakeShipAction {
            room: ShipRoom::Bridge,
            player_ix: 0,
        };

        let result = action.execute(&gs);
        assert!(result.is_ok());

        insta::with_settings!({sort_maps => true}, {
            // the tests here will force maps to sort
            insta::assert_json_snapshot!(result.unwrap());
        });
    }

    #[test_case(Sa(Some(Sas::default())); "Current in ship action")]
    #[test_case(Ep(None); "In event phase")]
    #[test_case(Cp {challenge: Challenge::default(), added: None}; "In challenge phase")]
    fn test_takeshipaction_err_if_wrong_phase(phase: GamePhase) {
        let gs = GameState::init_state().set_phase(phase).unwrap();

        let action = TakeShipAction {
            room: ShipRoom::Bridge,
            player_ix: 0,
        };

        let result = action.execute(&gs);
        assert!(result.is_err());
    }

    #[test_case(ShipRoom::Bridge)]
    #[test_case(ShipRoom::Deck)]
    #[test_case(ShipRoom::Galley)]
    fn test_takeshipaction_err_if_same_room(room: ShipRoom) {
        let gs = GameState::init_state().set_room(&room).unwrap();

        let action = TakeShipAction { room, player_ix: 0 };

        let result = action.execute(&gs);
        assert!(result.is_err());
    }

    #[test]
    fn test_deck_action() {
        let gs = GameState::init_state();

        let action = TakeShipAction {
            room: ShipRoom::Deck,
            player_ix: 0,
        };
        let result = action.execute(&gs);
        assert!(result.is_ok());

        insta::with_settings!({sort_maps => true}, {
            // the tests here will force maps to sort
            insta::assert_json_snapshot!(result.unwrap());
        });
    }

    #[test]
    fn test_galley_action() {
        let gs = GameState::init_state();

        let action = TakeShipAction {
            room: ShipRoom::Galley,
            player_ix: 0,
        };
        let result = action.execute(&gs);
        assert!(result.is_ok());

        insta::with_settings!({sort_maps => true}, {
            // the tests here will force maps to sort
            insta::assert_json_snapshot!(result.unwrap());
        });
    }
}
