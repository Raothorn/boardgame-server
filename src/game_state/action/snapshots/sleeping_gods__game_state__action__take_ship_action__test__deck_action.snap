---
source: src/game_state/action/take_ship_action.rs
expression: result.unwrap()
---
{
  "phase": {
    "ShipAction": {
      "DeckAction": {
        "search_tokens_drawn": []
      }
    }
  },
  "players": [
    {
      "command_tokens": 0,
      "hand": []
    }
  ],
  "crew": [
    {
      "name": "Sofi Odessa",
      "fatigue": 0,
      "skills": {
        "Craft": 2,
        "Savvy": 4
      }
    },
    {
      "name": "Laurant Lapointe",
      "fatigue": 0,
      "skills": {
        "Craft": 2,
        "Savvy": 4
      }
    }
  ],
  "ability_deck": {
    "items": [
      {
        "name": "1"
      },
      {
        "name": "2"
      },
      {
        "name": "3"
      },
      {
        "name": "4"
      },
      {
        "name": "5"
      },
      {
        "name": "6"
      },
      {
        "name": "7"
      },
      {
        "name": "8"
      },
      {
        "name": "9"
      }
    ],
    "discard": []
  },
  "search_token_deck": {
    "items": [
      1,
      2,
      3,
      4,
      5,
      6,
      7
    ],
    "discard": []
  },
  "event_card_deck": {
    "items": [
      {
        "name": "Biting Starfish",
        "options": [
          {
            "text": "Gain 1 Coin"
          },
          {
            "text": "Gain 1 Meat"
          }
        ]
      },
      {
        "name": "Broken Biplane",
        "options": [
          {
            "text": "Help repair the airplane: CRAFT 8"
          }
        ]
      }
    ],
    "discard": []
  },
  "room": "Deck",
  "resources": {
    "coins": 0,
    "grain": 0,
    "meat": 0
  },
  "prompt": {
    "promptData": {},
    "promptType": "drawForDeckAction"
  }
}
